core/client/db/src/storage_cache.rs

```diff
@@ -244,13 +251,22 @@ impl<H: Hasher, S: StateBackend<H>, B: Block> CachingState<H, S, B> {
 				cache.modifications.pop_back();
 			}
 			let mut modifications = HashSet::new();
-			for (k, v) in changes.into_iter() {
-				modifications.insert(k.clone());
-				if is_best {
-					cache.hashes.remove(&k);
-					CachingState::<H, S, B>::storage_insert(cache, k, v);
+			let childs = child_changes
+				.into_iter()
+				.map(|(k,i)|(Some(k),i))
+				.chain(::std::iter::once((None, changes)));
+			childs.for_each(|(sk, changes)|
+				for (k, v) in changes.into_iter() {
+					let k = (sk.clone(), k);
```

Comment:
This would need the tuple to derive borrow, which is probably not correct: relates to the idea of having a proper StorageKey type (and make this type such as it can implement borrow without instantiation (keeping an internal Vec<u8> with offset seems like the easiest way to me).

Reply:
Ok, my previous comment does probably not make much sense, HStorageKey is interesting but the problematic here is not that complicated.
The only simple non-break-all-api way of solving it I see is to put child in their own map storage. As said before it breaks lru expectation.
@tomusdrw I did notice that there is already two lru map, is it intended or is a single lru for hashes and decoded value a better choice?
So the way to fix this could be to use directly the internal lru storage (some `doublelinkedhashmap`) and manage lru limit size globally for `CachingState` (and of course move child cache to their own two level map). This lru size management does not looks to complicated and this will allow to have a single lru size.

Reply:
I'm not familiar with how caches are implemented, however it seems that currently child storage is used mostly for contracts, this means that having two separate lru's might make sense, cause we are switching between runtime context (that has it's own top-level cache) and contract context (that has it's own child cache). So after finishing the contract context we will still have all necessary top level items in the cache, no matter what contract did. I'm not sure if that's something we are going to keep though, if we do it might require a different solution.

Reply:
There is also the case of contract talking to each others, but I think you spot it: having separate mgmt could make sense (top level could be seen as more useful and child contract lru could use another size), plus it is more direct to implement :+1:

Reply:
Ok, I am not awake :man_facepalming: , there is already the kind of merged lru logic that I describe (we just got some overhead by using lru instead of its inner linkedhashmap struct) so I just need to keep using it that way https://github.com/paritytech/substrate/blob/69dd3b1b1646497360777880d3dfbcdd91a15fc4/core/client/db/src/storage_cache.rs#L58 and https://github.com/paritytech/substrate/blob/69dd3b1b1646497360777880d3dfbcdd91a15fc4/core/client/db/src/storage_cache.rs#L136